hazelcast-sql-core/src/test/java/com/hazelcast/sql/impl/expression/predicate/BetweenOperatorIntegrationTest.java

```diff
@@ -0,0 +1,235 @@
+        checkValues(sqlQuery("NOT BETWEEN CAST('08:00:00' AS TIME) AND CAST('20:00:00' AS TIME)"), TIME, new LocalTime[]{});
+        checkValues(sqlQuery("BETWEEN SYMMETRIC CAST('20:00:00' AS TIME) AND CAST('08:00:00' AS TIME)"), TIME, new LocalTime[]{time1, time2, time3});
+        checkValues(sqlQuery("BETWEEN SYMMETRIC CAST('08:00:00' AS TIME) AND CAST('20:00:00' AS TIME)"), TIME, new LocalTime[]{time1, time2, time3});
+        checkValues(sqlQuery("NOT BETWEEN SYMMETRIC CAST('20:00:00' AS TIME) AND CAST('08:00:00' AS TIME)"), TIME, new LocalTime[]{});
+    }
+
+    @Test
+    public void betweenPredicateTimestampsTest() {
+        LocalDateTime time1 = LocalDateTime.of(2000, 1, 1, 8, 0, 0);
+        LocalDateTime time2 = LocalDateTime.of(2010, 1, 1, 10, 0, 0);
+        LocalDateTime time3 = LocalDateTime.of(2020, 1, 1, 20, 0, 0);
+        putAll(time1, time2, time3);
+
+        checkValues(
+            sqlQuery("BETWEEN CAST('2000-01-01T08:00:00' AS TIMESTAMP) AND CAST('2020-01-01T20:00:00' AS TIMESTAMP)"),
+            TIMESTAMP,
+            new LocalDateTime[]{time1, time2, time3}
+        );
+        checkValues(
+            sqlQuery("BETWEEN CAST('2000-01-01T08:00:00' AS TIMESTAMP) AND CAST('2020-01-01T19:00:00' AS TIMESTAMP)"),
+            TIMESTAMP,
+            new LocalDateTime[]{time1, time2}
+        );
+        checkValues(
+            sqlQuery("BETWEEN CAST('2000-01-01T08:00:01' AS TIMESTAMP) AND CAST('2020-01-01T19:00:00' AS TIMESTAMP)"),
+            TIMESTAMP,
+            new LocalDateTime[]{time2})
+        ;
+        checkValues(
+            sqlQuery("NOT BETWEEN CAST('2000-01-01T08:00:00' AS TIMESTAMP) AND CAST('2020-01-01T20:00:00' AS TIMESTAMP)"),
+            TIMESTAMP,
+            new LocalDateTime[]{}
+        );
+        checkValues(
+            sqlQuery("BETWEEN SYMMETRIC CAST('2020-01-01T20:00:00' AS TIMESTAMP) AND CAST('2000-01-01T08:00:00' AS TIMESTAMP)"),
+            TIMESTAMP,
+            new LocalDateTime[]{time1, time2, time3}
+        );
+        checkValues(
+            sqlQuery("BETWEEN SYMMETRIC CAST('2000-01-01T08:00:00' AS TIMESTAMP) AND CAST('2020-01-01T20:00:00' AS TIMESTAMP)"),
+            TIMESTAMP,
+            new LocalDateTime[]{time1, time2, time3}
+        );
+        checkValues(
+            sqlQuery("NOT BETWEEN SYMMETRIC CAST('2020-01-01T20:00:00' AS TIMESTAMP) AND CAST('2000-01-01T08:00:00' AS TIMESTAMP)"),
+            TIMESTAMP,
+            new LocalDateTime[]{}
+        );
+    }
+
+    @Test
+    public void betweenPredicateDifferentTypesTest() {
+        LocalDate date = LocalDate.of(2000, 1, 1);
+        LocalTime time = LocalTime.of(8, 0, 0);
+        putAll(time, 1, "Argentina", "Bulgaria", "Ukraine", 2, date);
+
+        checkFailure0(sqlQuery("BETWEEN 1 AND 2"),
+            SqlErrorCode.DATA_EXCEPTION,
+            "Cannot parse VARCHAR value to BIGINT"
```

Comment:
We should also test cases that pass, e.g. when the VARCHAR can actually be parsed into a BIGINT.